4.8_ Grouping Data & Aggregating Variables

Task 01. Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise

In [3]:
# Import libraries 
import pandas as pd
import numpy as np
import os

In [5]:
# string
path =r'/Users/juliapetke/Desktop/Career Foundry/Instacart Basket Analysis'

In [97]:
# import of ords_prods_merge
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_new C.pkl'))

In [60]:
# Check dimensions of imported dataframe for consistency
ords_prods_merge.shape

(1000000, 19)

Task 02. In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe

In [62]:
ords_prods_merge.groupby('department_id')['order_number'].mean()

department_id
1     15.562844
2     17.518895
3     16.907152
4     17.537934
5     14.894521
6     16.658449
7     17.072954
8     14.942021
9     15.447580
10    18.681852
11    15.352278
12    14.327957
13    16.578690
14    16.666369
15    16.153766
16    17.768070
17    15.602964
18    19.674252
19    16.924344
20    16.292631
21    25.535596
Name: order_number, dtype: float64

Task 3. Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

Compared to the results from the subset, the results from the entire dataframe include all department_id values, ensuring a comprehensive representation of the data. Additionally, the mean values for the department_ids present in the subset have all changed, with some increasing and others decreasing.

Task 4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.


In [64]:
# Use groupby and transform() functions with np.max argument to determine maximum order number for each user
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/zd/kwmllg996knd4mqrwq8n_vf80000gp/T/ipykernel_54215/1602908596.py:2: FutureWarning: The provided callable <function max at 0x10562fec0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [70]:
df_ords_prods_merge.head (100)


product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
..         ...                         ...       ...            ...     ...   
95           1  Chocolate Sandwich Cookies        61             19     5.8   
96           1  Chocolate Sandwich Cookies        61             19     5.8   
97           1  Chocolate Sandwich Cookies        61             19     5.8   
98           1  Chocolate Sandwich Cookies        61             19     5.8   
99           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
..       ...      ...           ...                 ...                ...   
95    602103    10831             8                   3                 11   
96     49629    11119            18                   1                 14   
97    317888    11186            13                   5                 16   
98    682486    11243            16                   3                 13   
99   2861386    11448            10                   2                 15   

    days_since_prior_order  add_to_cart_order  reordered _merge  \
0                      3.0                  5          0   both   
1                     20.0                  1          1   both   
2                      6.0                 20          0   both   
3                      0.0                 10          0   both   
4                      9.0                 11          1   both   
..                     ...                ...        ...    ...   
95                    23.0                  5          0   both   
96                    30.0                  1          0   both   
97                     2.0                  8          0   both   
98                     0.0                  2          0   both   
99                     3.0                  6          0   both   

   merge_indicator        price_range     busiest_day    slowest_days  \
0             both  Mid-range product  Regularly busy  Regularly busy   
1             both  Mid-range product  Regularly busy  Regularly busy   
2             both  Mid-range product    Busiest days  Regularly busy   
3             both  Mid-range product  Regularly busy    Slowest days   
4             both  Mid-range product  Regularly busy    Slowest days   
..             ...                ...             ...             ...   
95            both  Mid-range product  Regularly busy    Slowest days   
96            both  Mid-range product    Busiest days  Regularly busy   
97            both  Mid-range product  Regularly busy  Regularly busy   
98            both  Mid-range product  Regularly busy    Slowest days   
99            both  Mid-range product  Regularly busy  Regularly busy   

   busiest_period_of_day  
0            Most orders  
1         Average orders  
2         Average orders  
3            Most orders  
4         Average orders  
..                   ...  
95           Most orders  
96           Most orders  
97           Most orders  
98           Most orders  
99           Most orders  

[100 rows x 19 columns]

In [72]:
# Use loc function to make loyalty flag column with given criteria
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [74]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [76]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [80]:
# Check frequency of values in new loyalty flag column
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    473041
Loyal customer      288151
New customer        238808
Name: count, dtype: int64

In [82]:
# View updated data frame
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  add_to_cart_order  reordered  _merge merge_indicator  \
0  ...                  5          0    both            both   
1  ...                  1          1    both            both   
2  ...                 20          0    both            both   
3  ...                 10          0    both            both   
4  ...                 11          1    both            both   

         price_range     busiest_day    slowest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regularly busy           Most orders   
1  Mid-range product  Regularly busy  Regularly busy        Average orders   
2  Mid-range product    Busiest days  Regularly busy        Average orders   
3  Mid-range product  Regularly busy    Slowest days           Most orders   
4  Mid-range product  Regularly busy    Slowest days        Average orders   

  max_order      loyalty_flag  
0        30  Regular customer  
1        30  Regular customer  
2         5      New customer  
3         3      New customer  
4         3      New customer  

[5 rows x 21 columns]

Task 5. The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [85]:
# Use groupby and agg() to determine basic statisitcs of spending habits by loyalty group
ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.781255  1.0  24.5
New customer      7.889919  1.0  24.5
Regular customer  7.859615  1.0  24.5

The average price of a product purchased by a new customer ($7.89) is slightly higher than that of a loyal customer ($7.78).

Task 6. The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [109]:
# Create a new column for the average spending of a customer
ords_prods_merge['average_spend'] = ords_prods_merge.groupby(['user_id'])['prices'].transform('mean')

In [113]:
 #Check the new column
ords_prods_merge.head(20)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
5    1701441      777            16                   1                  7   
6    1871483      825             3                   2                 14   
7    1290456      910            12                   3                 10   
8     369558     1052            10                   1                 20   
9     589712     1052            15                   1                 12   
10   1798115     1379             1                   5                 11   
11    497193     1480            36                   3                 13   
12   1727826     1480            38                   4                 13   
13   2694698     1480            39                   2                 12   
14   3171753     1494             6                   3                 10   
15    186761     1494             7                   2                  8   
16    317121     1494             9                   5                  8   
17    200390     1540             4                   2                  0   
18    374436     1540             5                   1                 21   
19   2920553     1540             6                   1                  0   

    days_since_prior_order  add_to_cart_order  reordered _merge  \
0                      3.0                  5          0   both   
1                     20.0                  1          1   both   
2                      6.0                 20          0   both   
3                      0.0                 10          0   both   
4                      9.0                 11          1   both   
5                     26.0                  7          0   both   
6                     30.0                  2          0   both   
7                     30.0                  1          0   both   
8                     19.0                  1          0   both   
9                     15.0         

In [129]:
# Assign spender tag to each customer
ords_prods_merge.loc[ords_prods_merge['average_spend']>= 10, 'spender_flag'] = 'High_spender'

In [131]:
ords_prods_merge.loc[ords_prods_merge['average_spend']<10,'spender_flag']='Low_spender'

In [133]:
# Check spender flags
ords_prods_merge['spender_flag'].value_counts(dropna = False)

spender_flag
Low_spender     791539
High_spender    208461
Name: count, dtype: int64

The result indicates that there are 791,539 low spenders and 208,461 high spenders in the customer base. This shows that the majority of customers are low spenders, while a smaller group accounts for higher spending. 

Task 7. In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:

If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [137]:
# Create a new column for customer frequency flags based on the median values from 'days_since_prior_order' column
ords_prods_merge['customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [139]:
# Check that the column was created successfully and properly
ords_prods_merge.head(20)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0    3139998      138            28                   6                 11   
1    1977647      138            30                   6                 17   
2     389851      709             2                   0                 21   
3     652770      764             1                   3                 13   
4    1813452      764             3                   4                 17   
5    1701441      777            16                   1                  7   
6    1871483      825             3                   2                 14   
7    1290456      910            12                   3                 10   
8     369558     1052            10                   1                 20   
9     589712     1052            15                   1                 12   
10   1798115     1379             1                   5                 11   
11    497193     1480            36                   3                 13   
12   1727826     1480            38                   4                 13   
13   2694698     1480            39                   2                 12   
14   3171753     1494             6                   3                 10   
15    186761     1494             7                   2                  8   
16    317121     1494             9                   5                  8   
17    200390     1540             4                   2                  0   
18    374436     1540             5                   1                 21   
19   2920553     1540             6                   1                  0   

    ...  reordered  _merge  merge_indicator        price_range  \
0   ...          0    both             both  Mid-range product   
1   ...          1    both             both  Mid-range product   
2   ...          0    both             both  Mid-range product   
3   ...          0    both             both  Mid-range product   
4   ...          1    both             both  Mid-range product   
5   ...          0    both             both  Mid-range product   
6   ...          0    both             both  Mid-range product   
7   ...          0    both             both  Mid-range product   
8   ...          0    both             both  Mid-range product   
9   ...          1    both             both  

In [141]:
# Based on the criteria provided, create a column featuring order frequency flags
ords_prods_merge.loc[ords_prods_merge['customer_frequency'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [143]:
ords_prods_merge.loc[(ords_prods_merge['customer_frequency'] > 10) & (ords_prods_merge['customer_frequency'] <=20), 'frequency_flag'] = 'Regular customer'

In [145]:
ords_prods_merge.loc[ords_prods_merge['customer_frequency'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [147]:
# Check value counts for new column
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent customer        684160
Regular customer         213442
Non-frequent customer    102398
Name: count, dtype: int64

In [149]:
ords_prods_merge[['user_id', 'customer_frequency', 'frequency_flag']].head(50)

user_id  customer_frequency         frequency_flag
0       138                 6.0      Frequent customer
1       138                 6.0      Frequent customer
2       709                 6.0      Frequent customer
3       764                 4.5      Frequent customer
4       764                 4.5      Frequent customer
5       777                26.0  Non-frequent customer
6       825                30.0  Non-frequent customer
7       910                30.0  Non-frequent customer
8      1052                12.0       Regular customer
9      1052                12.0       Regular customer
10     1379                 6.5      Frequent customer
11     1480                 7.0      Frequent customer
12     1480                 7.0      Frequent customer
13     1480                 7.0      Frequent customer
14     1494                 2.0      Frequent customer
15     1494                 2.0      Frequent customer
16     1494                 2.0      Frequent customer
17     1540                 7.0      Frequent customer
18     1540                 7.0      Frequent customer
19     1540                 7.0      Frequent customer
20     1540                 7.0      Frequent customer
21     1540                 7.0      Frequent customer
22     1540                 7.0      Frequent customer
23     1540                 7.0      Frequent customer
24     1540                 7.0      Frequent customer
25     1540                 7.0      Frequent customer
26     1540                 7.0      Frequent customer
27     1540                 7.0      Frequent customer
28     1540                 7.0      Frequent customer
29     1540                 7.0      Frequent customer
30     1540                 7.0      Frequent customer
31     1540                 7.0      Frequent customer
32     1540                 7.0      Frequent customer
33     1540                 7.0      Frequent customer
34     1598                 2.0      Frequent customer
35     1598                 2.0      Frequent customer
36     1598                 2.0      Frequent customer
37     1647                18.0       Regular customer
38     1647                18.0       Regular customer
39     1647                18.0       Regular customer
40     1647                18.0       Regular customer
41     1647                18.0       Regular customer
42     2833                 0.0      Frequent customer
43     2850                 9.0      Frequent customer
44     2857                10.0      Frequent customer
45     3010                 7.0      Frequent customer
46     3029                15.0       Regular customer
47     3395                30.0  Non-frequent customer
48     3904                 0.0      Frequent customer
49     4006                 7.0      Frequent customer

Task 9. Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [154]:
# Check shape before exporting
ords_prods_merge.shape

(1000000, 23)

In [152]:
# Export the updated dataframe as a pickle file
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined_merged_new_variables_4.8.pkl'))